operator/internal/controller/redpanda/index.go

```diff
@@ -11,55 +11,70 @@ package redpanda
 
 import (
 	"context"
+	"fmt"
+	"reflect"
 	"slices"
 
 	redpandav1alpha2 "github.com/redpanda-data/redpanda-operator/operator/api/redpanda/v1alpha2"
+	"github.com/redpanda-data/redpanda-operator/operator/pkg/functional"
 	appsv1 "k8s.io/api/apps/v1"
 	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/utils/ptr"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-const (
-	userClusterIndex        = "__user_referencing_cluster"
-	deploymentClusterIndex  = "__deployment_referencing_cluster"
-	statefulsetClusterIndex = "__statefulset_referencing_cluster"
-)
+type clientList[T client.Object] interface {
+	client.ObjectList
+	GetItems() []T
+}
 
-func userCluster(user *redpandav1alpha2.User) types.NamespacedName {
-	return types.NamespacedName{Namespace: user.Namespace, Name: user.Spec.ClusterSource.ClusterRef.Name}
+func clusterReferenceIndexName(name string) string {
+	return fmt.Sprintf("__%s_referencing_cluster", name)
 }
 
-func registerUserClusterIndex(ctx context.Context, mgr ctrl.Manager) error {
-	return mgr.GetFieldIndexer().IndexField(ctx, &redpandav1alpha2.User{}, userClusterIndex, indexUserCluster)
+func registerClusterSourceIndex[T client.Object, U clientList[T]](ctx context.Context, mgr ctrl.Manager, name string, o T, l U) (handler.EventHandler, error) {
+	indexName := clusterReferenceIndexName(name)
+	if err := mgr.GetFieldIndexer().IndexField(ctx, o, indexName, indexByClusterSource); err != nil {
+		return nil, err
+	}
+	return enqueueFromSourceCluster(mgr, name, l), nil
```

Comment:
So this is actually a bit of the opposite, but potentially yes.

The idea is that we trigger a reconciliation on the `Topic`/`User` based off of a change to the underlying `Redpanda` resource. So, let's say you change some configuration at the cluster level, we'll trigger reconciliation on all of the users, etc. (Not reconciliation of the cluster if a User/Topic changes).

Currently I have that on the `User` resource b/c I can envision cases where you have something like a cluster that doesn't have SCRAM-based auth enabled. If a User tries to be created with SCRAM creds, you're going to fail and the resource is going to be marked as `Synced = False`. If SCRAM auth is then enabled on the cluster, you'll want to re-reconcile the `User` resource since it can now be successfully synced.